// azure-iothub/dist/pl/models/bulkRegistryMappers.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.X509Thumbprint = exports.SymmetricKey = exports.PropertyContainer = exports.ExportImportDevice = exports.DeviceRegistryOperationWarning = exports.DeviceRegistryOperationError = exports.DeviceCapabilities = exports.BulkRegistryOperationResult = exports.AuthenticationMechanism = void 0;
var mappers_1 = require("../models/mappers");
Object.defineProperty(exports, "AuthenticationMechanism", { enumerable: true, get: function () { return mappers_1.AuthenticationMechanism; } });
Object.defineProperty(exports, "BulkRegistryOperationResult", { enumerable: true, get: function () { return mappers_1.BulkRegistryOperationResult; } });
Object.defineProperty(exports, "DeviceCapabilities", { enumerable: true, get: function () { return mappers_1.DeviceCapabilities; } });
Object.defineProperty(exports, "DeviceRegistryOperationError", { enumerable: true, get: function () { return mappers_1.DeviceRegistryOperationError; } });
Object.defineProperty(exports, "DeviceRegistryOperationWarning", { enumerable: true, get: function () { return mappers_1.DeviceRegistryOperationWarning; } });
Object.defineProperty(exports, "ExportImportDevice", { enumerable: true, get: function () { return mappers_1.ExportImportDevice; } });
Object.defineProperty(exports, "PropertyContainer", { enumerable: true, get: function () { return mappers_1.PropertyContainer; } });
Object.defineProperty(exports, "SymmetricKey", { enumerable: true, get: function () { return mappers_1.SymmetricKey; } });
Object.defineProperty(exports, "X509Thumbprint", { enumerable: true, get: function () { return mappers_1.X509Thumbprint; } });
