// azure-iothub/dist/pl/models/jobsMappers.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.X509Thumbprint = exports.TwinProperties = exports.Twin = exports.QueryResult = exports.JobResponse = exports.JobRequest = exports.JobProperties = exports.DeviceJobStatistics = exports.DeviceCapabilities = exports.CloudToDeviceMethod = void 0;
var mappers_1 = require("../models/mappers");
Object.defineProperty(exports, "CloudToDeviceMethod", { enumerable: true, get: function () { return mappers_1.CloudToDeviceMethod; } });
Object.defineProperty(exports, "DeviceCapabilities", { enumerable: true, get: function () { return mappers_1.DeviceCapabilities; } });
Object.defineProperty(exports, "DeviceJobStatistics", { enumerable: true, get: function () { return mappers_1.DeviceJobStatistics; } });
Object.defineProperty(exports, "JobProperties", { enumerable: true, get: function () { return mappers_1.JobProperties; } });
Object.defineProperty(exports, "JobRequest", { enumerable: true, get: function () { return mappers_1.JobRequest; } });
Object.defineProperty(exports, "JobResponse", { enumerable: true, get: function () { return mappers_1.JobResponse; } });
Object.defineProperty(exports, "QueryResult", { enumerable: true, get: function () { return mappers_1.QueryResult; } });
Object.defineProperty(exports, "Twin", { enumerable: true, get: function () { return mappers_1.Twin; } });
Object.defineProperty(exports, "TwinProperties", { enumerable: true, get: function () { return mappers_1.TwinProperties; } });
Object.defineProperty(exports, "X509Thumbprint", { enumerable: true, get: function () { return mappers_1.X509Thumbprint; } });
