// azure-iothub/dist/pl/models/queryMappers.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.X509Thumbprint = exports.TwinProperties = exports.Twin = exports.QuerySpecification = exports.QueryGetTwinsHeaders = exports.DeviceCapabilities = void 0;
var mappers_1 = require("../models/mappers");
Object.defineProperty(exports, "DeviceCapabilities", { enumerable: true, get: function () { return mappers_1.DeviceCapabilities; } });
Object.defineProperty(exports, "QueryGetTwinsHeaders", { enumerable: true, get: function () { return mappers_1.QueryGetTwinsHeaders; } });
Object.defineProperty(exports, "QuerySpecification", { enumerable: true, get: function () { return mappers_1.QuerySpecification; } });
Object.defineProperty(exports, "Twin", { enumerable: true, get: function () { return mappers_1.Twin; } });
Object.defineProperty(exports, "TwinProperties", { enumerable: true, get: function () { return mappers_1.TwinProperties; } });
Object.defineProperty(exports, "X509Thumbprint", { enumerable: true, get: function () { return mappers_1.X509Thumbprint; } });
